### ISSUE TITLE
book paths mangled

### ISSUE THREAD

**Issue by T-o-m-H-u** [noembryo/KoHighlights#31]
I wanted to try KoHighlights, but… things went wrong with the book pathnames.

- koreader on inkpad with homedir set to `/mnt/ext1`
- files and sidecars on inkpad in `/mnt/ext1/news`
- homedir exported through _sshfs_ as `/mnt/ext1`
- _sshfs_-mounted `inkpad:/mnt/ext1` to  `linux-host:$HOME/nfs/koreader`
- KoHighlights-1.7.3.0
- KoHighlights.v1.7.3.0.PySide2
- KoHighlights started by cli  
  `python $HOME/.local/bin/KoHighlights-1.7.3.0/main.py $HOME/nfs/koreader/news/`

The books’ paths are cut to an impossible absolute destination. I would expect the path to be relative to some known directory, e.g. the CWD, or an absolute system path, where it was last found (not ideal)
![kohighlight](https://github.com/noembryo/KoHighlights/assets/98460455/b6eb9383-c370-441e-abe8-1192cff3b292)

The paths to the sidecar files are correct – being absolute system paths perhaps also not ideal.
![luapaths](https://github.com/noembryo/KoHighlights/assets/98460455/8443ce4b-7ed0-44b5-814f-8d52439cd128)

It seems, KoHighlights cuts away the Home folder from the `doc_path` in the lua file, in my case: `/mnt/ext1`. How it knows about the homedir of koreader, escapes me. I would expect the basename to be relative to the input path given to KoHighlights.

- [x] When _news_ directory is rsync’ed to local FS, the book paths do not change.
- [x] When the mount is changed to `/mnt/ext1/news $HOME/nfs/koreader` the paths do not change either. The only thing koreader’s _dropbear-ssh-plugin_ allows to export, is the home-dir of koreader.

**Comment by noembryo** [comment]
You have to excuse my really basic knowledge of Linux, but I can't seem to understand the issue here..
Is it the appearance of the paths in the path column or is it something actually not working?
Are the paths working?
If you double-click a row, does the book opens?

If you like a different formatting of the paths (and you can modify the main.py file) you can change the # 945 from
`path_item = QTableWidgetItem(meta_path)` to
`path_item = QTableWidgetItem(abspath(meta_path))` or
`path_item = QTableWidgetItem(os.path.relpath(meta_path))` or
`path_item = QTableWidgetItem(os.path.realpath(meta_path))`
and see what looks better..
I don't use a Linux distribution but if there is something not working, I can check it in a VMWare image (that I keep for compiling).
The problem as I said, is that I know next to nothing about this OS..

If I totally missed your point, I'm sorry but you have to ELI5.. 😉

**Comment by T-o-m-H-u** [comment]
Sorry for being ambiguous. I know next to nothing about python. KoHighlights actually cannot find the files. It’s not a cosmetic issue.

The Book path in the list indicates the problem. It cannot possibly reside at `/news/document.pdf`. The news-directory is not located at `/` The correct and _absolute_ path is `/home/tom/nfs/koreader/news/document.pdf`.

The beginning of the Unix File System Hierarchy is the `/`. Every pathname that has leading slash starts at that root of filesystem. (That’s why we call it root, not slash) It’s strongly discouraged to litter the root with user files. There’s no equivalent in Win, because storage devices are mounted to drive letters. The closest equivalent would probably be the position before the drive letter, which has no name.
[a very short introduction to the Unix File System Hierarchy at eecs.WSU.edu](https://eecs.wsu.edu/~cs207/hierarchy.htm)

The problem probably results from the sdr file’s:`["doc_path"] = "/mnt/ext1/news/document.pdf"`. the _doc_path_ is extracted and somehow _/mnt/ext1_ is cut from it.

If the sidecar resides in the same directory as the document, the complete path should be cut until `document.pdf`, and then be reconstrcuted in conjunction with the directory where the sidecar was found.

I don’t understand how KoHighlights finds the books, that’s why this is all conjecture.

**Comment by noembryo** [comment]
OK. We have a problem.. 💯 
Played with it a little and I think fixed it.

Use the attached version to see if its working and tell me if its working so I can prepare an updated version..
(P2 is for Python2/PySide and P3 is for Python3/PySide2) 
[KOHighlights P3.zip](https://github.com/noembryo/KoHighlights/files/13919510/KOHighlights.P3.zip)
[KOHighlights P2.zip](https://github.com/noembryo/KoHighlights/files/13919514/KOHighlights.P2.zip)

**Comment by noembryo** [comment]
The problem with paths starts if the sidecar doesn't resides in the same directory as the document.
There are some cases that the metadata are inside a KOReader's History folder or in [`hashbasedstorage`](https://github.com/koreader/koreader/pull/10945#issuecomment-1742962415).
That's why I was using the ["doc_path"] as the first choice for book's path.
Now I use it only if the book is not at the same folder as the sidecar folder..

**Comment by T-o-m-H-u** [comment]
It works, thank you! The paths are now absolute. I hope it doesn’t make the files im-movable.

On the inkpad, I keep the sidecars in the same dir as the files. If one wants to keep them centrally, `doc_path` would probably still be a good starting point.

The hash-based-storage seems tricky as pdf checksums, as koreader notes, might change frequently with highlighting.

### archiving error
However, archiving throws the following error:
```
Traceback (most recent call last):
  File "/home/tom/.local/bin/KOHighlights P3/main.py", line 841, in on_archive
    data["stats"]["performance_in_pages"] = {}  # can be cluttered
    ~~~~^^^^^^^^^
KeyError: 'stats'
```
It seems related to koreader’s Reading statistics plugin https://github.com/koreader/koreader/issues/2986 Is this strictly needed? It’s currently disabled here, with no corresponding metadata in the sdr-lua-file.

### metadata not read

Metadata is not extracted. All titles equal `metadata.pdf` or `metadata.epub` and all authors are `OLD TYPE FILE`.

Exiftool and pdfinfo print metadata correctly, though. I realize pdf-metadata is convoluted. Is there a metadata type that is guranteed to be read, so it may be tested against?

**Comment by noembryo** [comment]
> However, archiving throws the following error:

You can check your `metadata.xxx.lua` to see if there is a `["stats"]` key.
You might be right about the statistics plugin.
I will guard against errors like this, you don't have to enable the plugin.

>metadata not read
Metadata is not extracted. All titles equal metadata.pdf or metadata.epub and all authors are OLD TYPE FILE.

Well, I don't actually extract metadata from the files..
Just reading the `metadata.xxx.lua` file.
If there is no author/title/.. there, I don't have them either..
.. or are you saying that this info exists in the `.lua` file and KOHighlights can't read it..?

Try this newer version..
[KOHighlights P3.zip](https://github.com/noembryo/KoHighlights/files/13921836/KOHighlights.P3.zip)

**Comment by T-o-m-H-u** [comment]
Thank you! This seems to wrap it up.

With the update the error regarding _stats_ was replaced by an identical one regarding _title_. It seems KoHighlights needs a title to archive highlights.

With the _Reading statistics_ plugin enabled (which writes the metadata to the sidecar, where KoHighlights can pick it up) title and author become available and archiving works.

So at the moment the stats plugin is necessary, unless KoHighlights would construct _title_ and _author_ by itself.

**Comment by noembryo** [comment]
> So at the moment the stats plugin is necessary, unless KoHighlights would construct title and author by itself.

Hmm.. I have to add this at [prerequisites](https://github.com/noembryo/KoHighlights#prerequisites) then..
BTW, by stats do you mean the System statistics plugin?

**Comment by T-o-m-H-u** [comment]
Sorry, I only meant the _Reading statistics_ plugins by 'stats'.

The  _Systems statistics_ plugin is disabled here. It seems not to be necessary – by I don’t know, what it does.

**Comment by noembryo** [comment]
OK
Thank you very much for your help 👍

**Comment by noembryo** [comment]
Fixed with v1.7.3.0
